### ISSUE TITLE
Allow stale configs

### ISSUE THREAD

**Issue by danielcompton** [envkey/envkey-source#5]
It would be nice if it was possible to allow stale configs while connected to the network. The use case is that I'm running envkey on my development machines when I enter the project directory, and 99.99% of the time there will be no change. This adds a delay of a second or two each time. I can see a few possible options:

* Add a config to allow stale caching up to some amount of time (say one hour)
* Add a config to allow [stale-while-revalidate](https://tools.ietf.org/html/rfc5861) behaviour for an additional amount of time, i.e. if cached data is less than 30 minutes old, use it, if it is between 30-60 minutes old then use it, but in the background update the cached data, and if it is over 60 minutes block and fetch the most recent data.

**Comment by danenania** [comment]
This is interesting. I've been considering a different approach to caching in the core [envkey-fetch](https://github.com/envkey/envkey-fetch) lib that would simplify this. Instead of caching to disk, a background process would start that caches in memory and then listens on a socket for updates. Apart from helping to minimize network calls and latency, there are security benefits, and it's also good for a process-per-request language like PHP.

I'll keep you posted on this!

**Comment by miquelbrazil** [comment]
I'd be very interested in this. I'm a PHP developer as well and I generally don't deploy to environments that would make setting persistent environment variables a great option. Instead, environment variables are set per request which usually means the presence of a `.env` in the project directory. Although restrictive permissions are set on the file, I realize this practice if frowned upon – but since envkey creates an encrypted cached copy of the variables, that could be left in place and my app could leverage that through a `shellexec` command to the envkey.

Maybe there is a workflow already for this use case.

**Comment by danenania** [comment]
@miquelbrazil Yes, php's process model can make using envkey tricky, especially if you don't have any access to the underlying os/shell like with shared hosting. Given the current architecture, I worry that building in time-based cache invalidation could create new problems, since it would effectively mean going from strong consistency to eventual consistency, with the unpredictability and edge cases that entails. That said, we are still planning to move to the background process approach I described above (I know it's been awhile), and that's the main blocker to creating an envkey php library where you wouldn't have to worry about any of this.

In the meantime, if you're ok with using `shellexec`, it shouldn't be too hard to hack around this yourself. You could pipe the `envkey-source` output to a `.env` file like this: `envkey-source YOUR-ENVKEY > .cached.env` and use a `dotenv` library to pull it in. To handle time-based invalidation, you could tack a timestamp variable onto the .env file and refresh periodically based on that.

**Comment by miquelbrazil** [comment]
Hey @danenania, thanks for such a fast response. If I didn't mention it, I love envkey and it's game changer for the projects I work on professionally and personally – I've been recommending it every chance I get.

I'm actually doing something similar to the approach you outlined. The credentials for the main project I'm using EnvKey with at the moment don't change quite that often so I just leave the `.env.{environment}` file in place and only fetch if it isn't detected. Newly fetched data is written back out to a new `.env` file using the `--dot-env-compatible` flag.

What I was trying to work around is what seems to be most people's admonishment of the security issues caused by a `.env` file in the production environment. I'm a bit skeptical of the argument because outside of a containerized environment it feels to me that persisting app configuration as environment variables, especially in shared hosting environments or even virtual hosted applications pollutes the namespace. I get around this by prefixing my config variables with a prefix defined at the app level – but even if I loaded them as environment variables, they only last as long as the request/response cycle. So I'm not sure I know any other way around having an `.env` file present. When I realize that the cached file that envkey-source created was encrypted, I figured maybe having an encrypted version of the `.env` file persist on disk that was decrypted on the fly would mitigate some of the security concerns (which I still think are minimal). But I also don't want to write a whole encrypt/decrypt implementation.

**Comment by danenania** [comment]
Thanks @miquelbrazil, that is awesome to hear!

On security, I think it comes down to your trust model. If an attacker has access to the os environment or file system, that almost always means that you're compromised no matter what approach you use to inject secrets. Trying to hide secrets from an attacker with this much access (while still making them accessible to your app) is effectively hopeless. So to me it makes the most sense to simply treat both the file system and environment as trusted and proceed accordingly.

**Comment by danenania** [comment]
In EnvKey v2, which has now launched, this can now be accomplished with the `--mem-cache / -m` flag, which keeps latest environment updated in RAM and doesn't make network requests on subsequent calls.

https://github.com/envkey/envkey

More on envkey-source v2, including this flag and other new features:

https://docs-v2.envkey.com/docs/envkey-source
